File: src/main/org/nlogo/api/Dump.java
package org.nlogo.api ;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

public final strictfp class Dump
{

	// this class is not instantiable
	private Dump() { throw new IllegalStateException() ; }

	///

	public static final CSV csv =
		new CSV
		( new CSV.ObjectDumper() {
				public String dump( Object obj ) {
					// We don't allow Integers anymore as Logo values,
					// but it's convenient to be able to dump them in
					// a CSV context... - ST 5/30/06
					if( obj instanceof Integer )
					{
						return obj.toString() ;
					}
					else
					{
						return Dump.logoObject( obj , true , true ) ;
					}
				} } ) ;

	///

	public static boolean isKnownType( Object obj )
	{
		return ( obj instanceof ExtensionObject ||
				 obj instanceof Boolean ||
				 obj instanceof Double ||
				 obj instanceof String ||
				 obj instanceof AgentSet ||
				 obj instanceof Agent ||
				 obj instanceof Nobody ||
				 obj instanceof List ) ;
	}

	public static String logoObject( Object obj )
	{
		return logoObject( obj, false, false ) ;
	}

	public static String logoObject( Object obj , boolean readable , boolean exporting )
	{
		// We need to check this first, otherwise those who subclass from the base types
		// when defining an ExtensionObject will never have their dump(...) called
		if( obj instanceof ExtensionObject )
		{
			// note that unless we directly call Dump.extensionObject we'll always be calling
			// reference = exporting.  I think that works since only the extension itself should
			// be calling !reference ev 2/29/08 
			return extensionObject( (org.nlogo.api.ExtensionObject) obj , readable , exporting , exporting ) ;
		}
		else if( obj instanceof Integer )
		{
			throw new IllegalArgumentException( "Integer: " + obj ) ;
		}
		else if( obj instanceof Boolean )
		{
			return obj.toString() ;
		}
		else if( obj instanceof Double )
		{
			return number( (Double) obj ) ;
		}
		else if( obj instanceof String )
		{
			if( readable )
			{
				return "\"" + StringUtils.escapeString( (String) obj ) + "\"" ;
			}
			else
			{
				return (String) obj ;
			}
		}
		else if( obj instanceof AgentSet )
		{
			return agentset( (AgentSet) obj, exporting ) ;
		}
		else if( obj instanceof Agent )
		{
			return agent( (Agent) obj, exporting ) ;
		}
		else if( obj instanceof Nobody )
		{
			return "nobody" ;
		}
		else if( obj instanceof List )
		{
			return list( (List<?>) obj , readable, exporting ) ;
		}
		else if( obj instanceof Lambda )
		{
			return obj.toString() ;
		}
		else if( obj == null )
		{
			return "<null>" ;
		}
		else
		{
			return "<" + obj.getClass().getName() + ">" ; 
		}
	}	

	public static String extensionObject( ExtensionObject obj, boolean readable, 
										  boolean exporting , boolean reference )
	{
		// The #{extension:type DATA}# format is treated as a literal when tokenizing
		return "{{" + obj.getExtensionName() + ":" + obj.getNLTypeName() + " "
			+ obj.dump( readable, exporting , reference )
			+"}}";
	}
	
	public static String number( Double obj )
	{
		// If there is some more efficient way to test
		// whether a double has no fractional part and
		// lies in IEEE 754's exactly representable range,
		// I would love to know about it. - ST 5/31/06
		double d = obj.doubleValue() ;
		long l = (long) d ;
		return
			( l == d &&
			  l >= -9007199254740992L &&
			  l <= 9007199254740992L )
			? Long.toString( l )
			: Double.toString( d ) ;
	}

	public static String number( double d )
	{
		long l = (long) d ;
		return
			( l == d &&
			  l >= -9007199254740992L &&
			  l <= 9007199254740992L )
			? Long.toString( l )
			: Double.toString( d ) ;
	}

	public static String map( Map<?,?> map )
	{
		return map( map , false, false ) ;
	}
	
	public static String map( Map<?,?> map , boolean readable, boolean exporting)
	{
		StringBuilder buff=new StringBuilder();
		for (Iterator<?> e = map.keySet().iterator() ; e.hasNext() ;)
		{
			String key=(String)e.next();
			//buff.append(key + " = " + hm.get(key) + "\n");
			Object value = map.get(key);
			if(value == null)
			{
				buff.append(key + " = null\n");
			}
			else if(value instanceof AgentSet)
			{
				buff.append(key + " = \n" + agentset( (AgentSet) value, exporting ));
			}
			else if(value instanceof Agent)
			{
				buff.append(key + " = \n" + agent( (Agent) value, exporting ));
			}
			else if(value instanceof List)
			{
				buff.append(key + " = \n" + list( (List<?>) value, readable, exporting ));
			}
			else
			{
				buff.append(key + " = " + value + "\n");
			}
		}
		return buff.toString() ;
	}
	
	public static String list( List<?> list )
	{
		return list( list, false, false ) ;
	}

	public static String list( List<?> list, boolean readable, boolean exporting )
	{
		return iterator( list.iterator(), readable , exporting ) ;
	}

	public static String list( List<?> list, String prefix, String suffix, String delimiter )
	{
		return iterator(list.iterator(), prefix, suffix, delimiter, false, false);
	}

	public static String list( List<?> list, String prefix, String suffix, String delimiter, boolean readable, boolean exporting )
	{
		return iterator(list.iterator(), prefix, suffix, delimiter, readable, exporting);
	}

	public static String iterator( Iterator<?> iter )
	{
		return iterator( iter , "[" , "]" , " " , false , false ) ;
	}

	public static String iterator(Iterator<?> iter, boolean readable, boolean exporting)
	{
		return iterator( iter , "[" , "]" , " " , readable, exporting ) ;
	}

	public static String iterator(Iterator<?> iter, String prefix, String suffix, String delimiter, boolean readable, boolean exporting)
	{
		StringBuilder buff=new StringBuilder();
		buff.append( prefix ) ;
		boolean firstOne = true ;
		while( iter.hasNext() )
		{
			if( firstOne )
			{
				firstOne = false ;
			}
			else
			{
				buff.append( delimiter ) ;
			}
			buff.append(logoObject(iter.next(), readable, exporting));
		}
		buff.append( suffix ) ;
		return buff.toString() ;
	}

	public static String agentset( AgentSet agentset, boolean exporting )
	{
		String printName = agentset.printName() ;
		if( printName != null )
		{
			printName = printName.toLowerCase() ;
		}
		StringBuilder buffer = new StringBuilder() ;
		if( !exporting )
		{
			if( printName != null )
			{
				return printName ;
			}
			buffer.append( "(agentset, " + agentset.count() + " " ) ;
			if( Turtle.class.isAssignableFrom( agentset.type() ) )
			{
				if( agentset.count() == 1 )
				{
					buffer.append( "turtle" ) ;
				}
				else
				{
					buffer.append( "turtles" ) ;
				}
			}
			else if( Patch.class.isAssignableFrom( agentset.type() ))
			{
				if( agentset.count() == 1 )
				{
					buffer.append( "patch" ) ;
				}
				else
				{
					buffer.append( "patches" ) ;
				}
			}
			else if( Observer.class.isAssignableFrom( agentset.type() ) )
			{
				buffer.append( "observer" ) ;
			}
			else if( Link.class.isAssignableFrom( agentset.type() ) )
			{
				if( agentset.count() == 1 )
				{
					buffer.append( "link" ) ;
				}
				else
				{
					buffer.append( "links" ) ;
				}				
			}
			else
			{
				throw new IllegalStateException() ;
			}
			buffer.append( ")" ) ;
		}
		else
		{
			buffer.append( "{");
			if( Turtle.class.isAssignableFrom( agentset.type() ) )
			{
				if( printName != null )
				{
					if( agentset == agentset.world().turtles() )
					{
						buffer.append( "all-" + printName ) ;
					}
					else
					{
						buffer.append( "breed " + printName ) ;
					}
				}
				else
				{
					buffer.append( "turtles" ) ;
					for( Agent a : agentset.agents() )
					{
						buffer.append( " " + ( (Turtle) a ).id() ) ;
					}
				}
			}
			else if( Link.class.isAssignableFrom( agentset.type() ) )
			{
				if( printName != null )
				{
					if( agentset == agentset.world().links() )
					{
						buffer.append( "all-" + printName ) ;
					}
					else
					{
						buffer.append( "breed " + printName ) ;
					}
				}
				else
				{
					buffer.append( "links" ) ;
					for( Agent a : agentset.agents() )
					{
						Link link = (Link) a ;
						buffer.append( " [" + link.end1().id() + " " +
									   link.end2().id() + " " +
									   agentset( link.getBreed() , true ) + "]" ) ;
					}
				}
			}
			else if( Patch.class.isAssignableFrom( agentset.type() ) )
			{
				if( printName != null )
				{
					buffer.append( "all-" + printName ) ;
				}
				else
				{
					buffer.append( "patches" ) ;
					for( Agent a : agentset.agents() )
					{
						Patch patch = (Patch) a ;
						buffer.append( " [" + patch.pxcor() + " " + patch.pycor() + "]") ;
					}
				}
			}
			else if( Observer.class.isAssignableFrom( agentset.type() ) )
			{
				buffer.append( "observer" ) ;
			}
			else
			{
				throw new IllegalStateException() ;
			}
			buffer.append( "}" ) ;
		}

		return buffer.toString() ;
	}

	public static String agent( Agent agent , boolean exporting )
	{
		String openParen = "(" ;
		String closeParen = ")" ;
		if( exporting )
		{
			openParen = "{" ;
			closeParen = "}" ;
		}
		if( agent instanceof Observer )
		{
			return "observer" ;
		}
		else if ( agent instanceof Turtle ||
				  agent instanceof Patch ||
				  agent instanceof Link )
		{			
			if( agent.id() == -1 )
			{
				return "nobody" ;
			}
			return openParen + agent.toString() + closeParen ;
		}
		throw new IllegalArgumentException( agent.toString() ) ;
	}

	public static String typeName( Object obj )
	{
		if ( obj instanceof Agent )
		{
			return typeName( obj.getClass() ) ;
		}
		else if( obj instanceof ExtensionObject )
		{
			return ( (ExtensionObject) obj).getNLTypeName();
		}
		else
		{
			return typeName( obj.getClass() ) ;
		}
	}

	public static String typeName( Class<?>theClass )
	{
		if( theClass == Boolean.class )
		{
			return "true/false" ;
		}
		else if( theClass == Double.class )
		{
			return "number" ;
		}
		else if( theClass == String.class )
		{
			return "string" ;
		}
		else if( AgentSet.class.isAssignableFrom( theClass ) )
		{
			return "agentset" ;
		}
		else if( Turtle.class.isAssignableFrom( theClass ) )
		{
			return "turtle" ;
		}
		else if( Patch.class.isAssignableFrom( theClass ) )
		{
			return "patch" ;
		}
		else if( Link.class.isAssignableFrom( theClass ) )
		{
			return "link" ;
		}
		else if( Observer.class.isAssignableFrom( theClass ) )
		{
			return "observer" ;
		}
		else if( Agent.class.isAssignableFrom( theClass ) )
		{
			return "agent" ;
		}
		else if( LogoList.class.isAssignableFrom( theClass ) )
		{
			return "list" ;
		}
		else if( Nobody.class.isAssignableFrom( theClass ) )
		{
			return "nobody" ;
		}
		else
		{
			return theClass.getName() ;
		}
	}
	
}

File: netlogo-gui/src/main/api/Dump.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.api

import collection.JavaConverters._

object Dump {

  private val dumper: Any => String = {
    // We don't allow Integers anymore as Logo values, but it's convenient to be able to dump them
    // in a CSV context. - ST 5/30/06
    case i: java.lang.Integer =>
      i.toString
    case x =>
      logoObject(x.asInstanceOf[AnyRef], true, true)
  }

  val csv = new CSV(dumper)

  def isKnownType(obj: AnyRef): Boolean =
    obj.isInstanceOf[java.lang.Boolean] ||
    obj.isInstanceOf[java.lang.Double] ||
    obj.isInstanceOf[String] ||
    obj.isInstanceOf[Agent] ||
    obj.isInstanceOf[AgentSet] ||
    obj.isInstanceOf[LogoList] ||
    obj.isInstanceOf[ExtensionObject] ||
    (obj eq Nobody)

  def logoObject(obj: AnyRef): String =
    logoObject(obj, false, false)

  def logoObject(obj: AnyRef, readable: Boolean, exporting: Boolean): String =
    obj match {
      // We need to check this first, otherwise those who subclass from the base types
      // when defining an ExtensionObject will never have their dump(...) called
      case eo: ExtensionObject =>
        // note that unless we directly call Dump.extensionObject we'll always be calling
        // reference = exporting.  I think that works since only the extension itself should
        // be calling !reference ev 2/29/08
        extensionObject(eo, readable, exporting, exporting)
      case i: java.lang.Integer =>
        sys.error("java.lang.Integer: " + i)
      case b: java.lang.Boolean =>
        b.toString
      case d: java.lang.Double =>
        number(d)
      case s: String =>
        if (readable)
          "\"" + StringUtils.escapeString(s) + "\""
        else s
      case a: AgentSet =>
        agentset(a, exporting)
      case a: Agent =>
        agent(a, exporting)
      case Nobody =>
        "nobody"
      case l: LogoList =>
        list(l, readable, exporting)
      case task: Task =>
        task.toString
      case null =>
        "<null>"
      case _ =>
        "<" + obj.getClass.getSimpleName + ">"
    }

  def extensionObject(obj: ExtensionObject, readable: Boolean,
                      exporting: Boolean, reference: Boolean) =
    // The #{extension:type DATA}# format is treated as a literal when tokenizing
    "{{" + obj.getExtensionName + ":" + obj.getNLTypeName + " " +
      obj.dump(readable, exporting, reference) + "}}"

  def number(obj: java.lang.Double) = {
    // If there is some more efficient way to test whether a double has no fractional part and lies
    // in IEEE 754's exactly representable range, I don't know it. - ST 5/31/06
    val d = obj.doubleValue
    val l = d.toLong
    if(l == d && l >= -9007199254740992L && l <= 9007199254740992L)
      l.toString
    else
      d.toString
  }

  def number(d: Double) = {
    val l = d.toLong
    if(l == d && l >= -9007199254740992L && l <= 9007199254740992L)
      l.toString
    else
      d.toString
  }

  def list(list: LogoList, readable: Boolean = false, exporting: Boolean = false) =
    iterator(list.scalaIterator, "[", "]", " ", readable, exporting)

  def iterator(iter: Iterator[AnyRef], prefix: String, suffix: String, delimiter: String, readable: Boolean, exporting: Boolean): String =
    iter.map(logoObject(_, readable, exporting))
      .mkString(prefix, delimiter, suffix)

  // This is fugly. But it really is this complicated - ST 6/27/11
  def agentset(as: AgentSet, exporting: Boolean): String =
    if (!exporting)
      Option(as.printName).map(_.toLowerCase).getOrElse{
        val buf = new StringBuilder
        buf ++= "(agentset, " + as.count + " "
        if (classOf[Turtle].isAssignableFrom(as.`type`)) {
          buf ++= "turtle"
          if (as.count != 1)
            buf += 's'
        }
        else if (classOf[Link].isAssignableFrom(as.`type`)) {
          buf ++= "link"
          if (as.count != 1)
            buf += 's'
        }
        else if (classOf[Patch].isAssignableFrom(as.`type`)) {
          buf ++= "patch"
          if (as.count != 1)
            buf ++= "es"
        }
        else if (classOf[Observer].isAssignableFrom(as.`type`))
          buf ++= "observer"
        else throw new IllegalStateException
        buf += ')'
        buf.toString
      }
    else {
      val buf = new StringBuilder
      buf += '{'
      if (classOf[Turtle].isAssignableFrom(as.`type`)) {
        val printName = as.printName
        if (printName != null) {
          buf ++=
            (if (as eq as.world.turtles)
               "all-"
             else
               "breed ")
          buf ++= printName.toLowerCase
        }
        else {
          val ids = as.agents.asScala.map(_.id.toString).toSeq
          buf ++= ("turtles" +: ids).mkString(" ")
        }
      }
      else if (classOf[Link].isAssignableFrom(as.`type`)) {
        val printName = as.printName
        if (printName != null) {
          buf ++=
            (if (as eq as.world.links)
               "all-"
             else
               "breed ")
          buf ++= printName.toLowerCase
        }
        else {
          def linkString(link: Link) =
            "[" + link.end1.id + " " + link.end2.id + " " + agentset(link.getBreed, true) + "]"
          val ids = as.agents.asScala.map(a => linkString(a.asInstanceOf[Link])).toSeq
          buf ++= ("links" +: ids).mkString(" ")
        }
      }
      else if (classOf[Patch].isAssignableFrom(as.`type`)) {
        val printName = as.printName
        if (printName != null) {
          buf ++= "all-"
          buf ++= printName.toLowerCase
        }
        else {
          def patchString(p: Patch) =
            "[" + p.pxcor + " " + p.pycor + "]"
          val ids = as.agents.asScala.map(a => patchString(a.asInstanceOf[Patch])).toSeq
          buf ++= ("patches" +: ids).mkString(" ")
        }
      }
      else if (classOf[Observer].isAssignableFrom(as.`type`))
        buf ++= "observer"
      else
        sys.error("unknown agentset type: " + as.`type`)
      buf.append("}")
      buf.toString
    }

  def agent(agent: Agent, exporting: Boolean) =
    if (agent.isInstanceOf[Observer])
      "observer"
    else if (agent.id == -1)
      "nobody"
    else {
      val (open, close) =
        if(exporting) ('{', '}')
        else ('(', ')')
      open + agent.toString + close
    }

  // not clear typeName really belongs here, but not sure where else to put it either - ST 6/27/11

  def typeName(obj: AnyRef): String =
    obj match {
      case eo: ExtensionObject =>
        eo.getNLTypeName
      case _ =>
        typeName(obj.getClass)
    }

  def typeName(clazz: Class[_]): String =
    clazz match {
      case _ if clazz eq classOf[java.lang.Boolean] =>
        "true/false"
      case _ if clazz eq classOf[java.lang.Double] =>
        "number"
      case _ if clazz eq classOf[String] =>
        "string"
      case _ if classOf[AgentSet].isAssignableFrom(clazz) =>
        "agentset"
      case _ if classOf[Turtle].isAssignableFrom(clazz) =>
        "turtle"
      case _ if classOf[Patch].isAssignableFrom(clazz) =>
        "patch"
      case _ if classOf[Link].isAssignableFrom(clazz) =>
        "link"
      case _ if classOf[Observer].isAssignableFrom(clazz) =>
        "observer"
      case _ if classOf[Agent].isAssignableFrom(clazz) =>
        "agent"
      case _ if classOf[LogoList].isAssignableFrom(clazz) =>
        "list"
      case _ if Nobody.getClass.isAssignableFrom(clazz) =>
        "nobody"
      case _ =>
        clazz.getSimpleName
    }

}
